<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-del-projecto-team_109/blob/main/Semana3_Actividad1_team_109.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Semana 3 - Actividad 1**

## José Antonio Mendoza Castro - A01794067
## Miguel Angel Aguilera Rodríguez - A00642541

# **Parte 1 - Fundamentos de bases de datos**

#### **1. Fundamentos de bases de datos y para ciencia de datos.**

Las bases de datos pueden ser guardadas de muchas maneras, principalmente de manera estructurada y no estructurada. La que conocemos de manera más habitual es la estructurada la cual se enfoca principalmente en formatos tabulares que contiene texto o números.
Las bases de datos relacionables más comunes de encontrar son en los sistemas:
*Oracle
*MySQL
*MS SQL Server
*PostgreSQL
*Amazon Redshift
*IBM DB2
*MS Access
*SQLite
*Snowflake

Una de las maneras en las que mejor se puede aprovechar la base de datos es conectándola directamente con código como Python, R o el propio SQL.


Es de suma importancia que antes de comenzar a analizar los datos, garantizar que esta se encuentre en las condiciones adecuadas.

El termino “limpieza de datos” hace énfasis en que la información que se mete a un modelo, si no es de calidad, el resultante no será de calidad. Por lo que el establecimiento de criterios en las informaciones para tener un gobierno de datos en las bases de datos es de suma importancia.



#### **2. Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos.**

Los almacenes de datos es un sistema de gestión el cual está diseñado para facilitar el análisis de la información recolectada y facilita la toma de decisiones del negocio.

Este almacenamiento de datos tiene como principal finalidad la centralización de información masiva. Es importante que estos almacenes sean custodiados para garantizar la integridad de la información ya que al ser la fuente de verdad absoluta muchas veces, al ser corrompida y no detectada, puede llevar a mala toma de decisiones.

# **Parte 2 - Selección y limpieza de los datos en Python**

#### Importar librerías

In [1]:
import pandas as pd
import numpy as np

#### Configurar opciones de Pandas para mostrar todas las columnas del *dataframe*

In [2]:
pd.set_option('display.max_columns', 100)

#### Lectura del *dataframe*

In [3]:
df = pd.read_csv('default of credit card clients.csv')

#### Visualización general de todos los campos a partir del método *.head()*

In [4]:
df.head()

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  X10  X11      X12  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0 -2.0 -2.0   3913.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  0.0  2.0   2682.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  0.0  0.0  29239.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  0.0  0.0  46990.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  0.0  0.0   8617.0   

       X13      X14      X15      X16      X17     X18      X19      X20  \
0   3102.0    689.0      0.0      0.0      0.0     0.0    689.0      0.0   
1   1725.0   2682.0   3272.0   3455.0   3261.0     0.0   1000.0   1000.0   
2  14027.0  13559.0  14331.0  14948.0  15549.0  1518.0   1500.0   1000.0   
3  48233.0  49291.0  28314.0  28959.0  29547.0  2000.0   2019.0   1200.0   
4   5670.0  35835.0  20940.0  19146.0  19131.0  2000.0  36681.0  10000.0   

      X21     X22     X23    Y  
0     0.0     0.0     0.0  1.0  
1  1000.0     0.0  2000.0  1.0  
2  1000.0  1000.0  5000.0  0.0  
3  1100.0  1069.0  1000.0  0.0  
4  9000.0   689.0   679.0  0.0

#### Descripción general de las variables

#### Lo que se aprecia al ver una primera descripción (aunque se necesitará revisar de manera individual cada variable):

1. Tenemos un *dataframe* con 29,984 filas

1.   X3 (Educación) tiene más valores a los que se describe en la información de atributos (1-4)

4. X4 (Estatus marital) tiene más valores a los que se describe en la información de atributos (1-3)

1.   X12-X17 (Importes de estado de cuenta) presentan datos negativos por lo que se interpretaría que hay clientes con sobrepago



Método *.describe()* en Pandas

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html

In [5]:
df.describe().round(1)

ID         X1       X2       X3       X4       X5       X6  \
count  30000.0    30000.0  29999.0  29998.0  29998.0  29995.0  29997.0   
mean   15000.5   167484.3      1.6      1.9      1.6     35.5     -0.0   
std     8660.4   129747.7      0.5      0.8      0.5      9.2      1.1   
min        1.0    10000.0      1.0      0.0      0.0     21.0     -2.0   
25%     7500.8    50000.0      1.0      1.0      1.0     28.0     -1.0   
50%    15000.5   140000.0      2.0      2.0      2.0     34.0      0.0   
75%    22500.2   240000.0      2.0      2.0      2.0     41.0      0.0   
max    30000.0  1000000.0      2.0      6.0      3.0     79.0      8.0   

            X7       X8       X9      X10      X11       X12       X13  \
count  29995.0  29993.0  29991.0  29984.0  29986.0   29989.0   29989.0   
mean      -0.1     -0.2     -0.2     -0.3     -0.3   51236.9   49190.7   
std        1.2      1.2      1.2      1.1      1.2   73645.2   71183.4   
min       -2.0     -2.0     -2.0     -2.0     -2.0 -165580.0  -69777.0   
25%       -1.0     -1.0     -1.0     -1.0     -1.0    3565.0    2986.0   
50%        0.0      0.0      0.0      0.0      0.0   22387.0   21207.0   
75%        0.0      0.0      0.0      0.0      0.0   67139.0   64027.0   
max        8.0      8.0      8.0      8.0      8.0  964511.0  983931.0   

             X14       X15       X16       X17       X18        X19       X20  \
count    29987.0   29985.0   29983.0   29990.0   29992.0    29991.0   29992.0   
mean     47025.4   43275.7   40324.5   38881.1    5662.9     5922.5    5225.6   
std      69360.9   64345.5   60810.0   59561.3   16564.2    23044.2   17608.4   
min    -157264.0 -170000.0  -81334.0 -339603.0       0.0        0.0       0.0   
25%       2667.5    2329.0    1763.5    1256.2    1000.0      835.5     390.0   
50%      20089.0   19052.0   18107.0   17081.0    2100.0     2009.0    1800.0   
75%      60182.0   54560.0   50213.0   49208.2    5006.0     5000.0    4505.5   
max    1664089.0  891586.0  927171.0  961664.0  873552.0  1684259.0  896040.0   

            X21       X22       X23        Y  
count   29989.0   29989.0   29995.0  29997.0  
mean     4827.3    4800.3    5216.3      0.2  
std     15668.8   15280.8   17778.8      0.4  
min         0.0       0.0       0.0      0.0  
25%       296.0     251.0     118.0      0.0  
50%      1500.0    1500.0    1500.0      0.0  
75%      4014.0    4033.0    4000.0      0.0  
max    621000.0  426529.0  528666.0      1.0

#### Exploración de *missing values*
#### Dado que tenemos 29,986 filas podemos eliminar los *missing values* o bien podemos imputar alguna medida como la media, mediana o moda sin afectar en un momento el desempeño de cualquier modelo que realicemos posteriormente, ya que son pocos los datos faltantes

Método *.isnull()* en Pandas

https://pandas.pydata.org/docs/reference/api/pandas.isnull.html

In [6]:
df.isnull().sum()

ID      0
X1      0
X2      1
X3      2
X4      2
X5      5
X6      3
X7      5
X8      7
X9      9
X10    16
X11    14
X12    11
X13    11
X14    13
X15    15
X16    17
X17    10
X18     8
X19     9
X20     8
X21    11
X22    11
X23     5
Y       3
dtype: int64

#### Hay que observar que de la variable X6 a la X11 tenemos el comportamiento de pago por lo que si bien podríamos imputar la moda no tendríamos el comportamiento real de esos clientes y la imputación no resultaría del todo correcta.

#### Las variables de X12 a la X23 se refieren a los montos de pago de manera histórica de cada cliente por lo que están relacionadas con las variables X6 a X11, con lo que imputar una media a los datos faltantes no sería correcto.

#### Dado lo anterior se eliminarán esas filas.

Método *.dropna()* en Pandas

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html

In [7]:
cols = df.columns[6:-1]

df.dropna(subset = cols, inplace = True)

#### Volvemos a verificar si aún tenemos datos nulos por lo que observamos que en la variable X5 (Edad) tenemos algunos datos nulos, ante lo cual imputaremos la media

In [8]:
df.isnull().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     3
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
X17    0
X18    0
X19    0
X20    0
X21    0
X22    0
X23    0
Y      0
dtype: int64

#### **X5 (Edad)**: se utilizará la media para cubrir los datos faltantes

In [9]:
x5_mean = df['X5'].mean()

Método *.fillna()* en Pandas

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

In [10]:
df['X5'] = df['X5'].fillna(x5_mean) 

### Se revisarán las variables X2 a X4 para ver si cumplen con las características establecidas en la descripción de atributos

#### **X2 (Género)**

In [11]:
df['X2'].unique()

array([2., 1.])

#### **X3 (Educación)**: los valores tendrían que estar entre 1 y 4, como observamos a continuación hay valores iguales a **0**, **5** y **6** los cuales serán remplazados con la moda

In [12]:
df['X3'].unique()

array([2., 1., 3., 5., 4., 6., 0.])

In [13]:
df.groupby('X3')['ID'].count()

X3
0.0       14
1.0    10574
2.0    14010
3.0     4909
4.0      123
5.0      280
6.0       51
Name: ID, dtype: int64

In [14]:
x3_mode = int(df['X3'].mode())

In [15]:
df['X3'] = df['X3'].astype(int)

In [16]:
df['X3'] = df['X3'].map({0: x3_mode, 1: 1, 2: 2, 3: 3, 4: 4, 5: x3_mode, 6: x3_mode})

In [17]:
df['X3'].unique()

array([2, 1, 3, 4])

#### **X4 (Estatus marital)**: Se tienen valores distintos a los definidos en los atributos, hay valores iguales a **0**, estos valores se reemplazarán con la moda

In [18]:
df['X4'].unique()

array([1., 2., 3., 0.])

In [19]:
df.groupby('X4')['ID'].count()

X4
0.0       54
1.0    13644
2.0    15940
3.0      323
Name: ID, dtype: int64

In [20]:
x4_mode = int(df['X4'].mode())

In [21]:
df['X4'] = df['X4'].astype(int)

In [22]:
df['X4'] = df['X4'].map({0: x4_mode, 1: 1, 2: 2, 3: 3})

In [23]:
df['X4'].unique()

array([1, 2, 3])

# **Parte 3 - Preparación de los datos**

#### **1. ¿Qué datos considero mas importantes? ¿Por qué?**

Los datos de comportamiento de pago, dado que están relacionados entre ellos no se deberían tratar con simples imputaciones de media, mediana y moda.

#### **2. ¿Se eliminaron o reemplazaron datos nulos? ¿Qué se hizo y por qué?**

Por una parte se eliminaron datos de las variables X6 a X23 dado que eran variables de comportamiento de pago. Se eliminaron dado que el número de *missing values* no es grande en la base de datos.

También se reemplazaron valores de la variable X3 a la X5, se ocupo la moda en X3 y X4, por otra parte la media en X5. Lo anterior en base a la descripción de atributos de la base de datos, por lo que se reemplazaron para que tuvieran el comportamiento deseado.

#### **3. ¿Es necesario limpiar los datos para el análisis? Sí / No / ¿Por qué?**

Sí, ya que la limpieza es un paso previo a modelar datos, es importante hacerlo por una parte porqué los modelos de Machine Learning no aceptan valores nulos y por otra parte para mejorar el desempeño del modelo que se realizará, es decir, para que en el caso que se haga una predicción, segmentación, etc. se pueda tomar una mejor decisión a partir de los resultados.

#### **4. ¿Existen problemas de formato que deban solucionar antes del proceso de modelado? Sí / No / Por qué.**

En la base de datos tenemos dos grupos de datos, por una parte variables categóricas (X2-X4, X6-X11) y variables numéricas (X5, X12-X23), por lo que en el caso de las variables categóricas deberían quedar como enteros y no como flotantes. 

#### **5. ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?**

Los ajustes que se realizaron fueron:

1. Quitar filas en el caso de atributos que no sería del todo correcto simplemente imputar alguna medida.

2. Reemplezar valores que salen fuera de las definiciones dadas en la descripción de los atributos.